Interpret data to normalize as ndarrays

ngoldbaum opened the pull request:
This follows on from #6622. That PR fixed the case where we were calling imshow with linear normalization. This fixes the case when we call imshow with log, symlog, or power normalization. For example:

```python
import numpy as np

from matplotlib import pyplot as plt
from matplotlib.colors import LogNorm
from yt.units import km

plt.imshow(np.random.random((16, 16))*km, norm=LogNorm())
```

Currently this dies with a unit error in LogNorm. Analogous to the fix applied in #6622 the fix is to interpret data handed to us by a user as an ndarray.

efiring commented:
Can this as well as the #6622 fix be consolidated into `Normalize.process_value()`?  That seems to me to be the logical place to strip off the units.

ngoldbaum commented:
Added logic to `process_value`. Unfortunately I could do that for all the normalizers since some of them do some additional data processing after the call to `process_value`.

I also added a test.

efiring commented on lib/matplotlib/colors.py:
```diff
@@ -907,8 +907,11 @@ def process_value(value):
+        # interpreted as an ndarray. See issue #6622.
         result = np.ma.array(value, dtype=dtype, copy=True)
-        return result, is_scalar
+        result_data = np.asarray(result.data)
```
I think that all you really need to do here is add the kwarg `subok=False` to the `np.ma.array()` line returning `result`.  Then `result` will be a plain masked array, and result.data will be a plain ndarray, so you will not need to extract and return it here.  The reason all this came up is that `np.ma.array` has a surprising default of `subok=True`, unlike `np.array`, which has the default `subok=False`.

ngoldbaum commented:
@efiring `subok` didn't seem to work (I think it only affects the type of `result`, not `result.data`) but the approach I just pushed does seem to work. Are you ok with just unconditionally calling np.asarray unless we're passed a masked array?

efiring commented:
Try this substitution:
```python
# result = np.ma.array(value, dtype=dtype, copy=True)
mask = np.ma.getmask(value)
data = np.asarray(np.ma.getdata(value))
result = np.ma.array(data, mask=mask, dtype=dtype, copy=True)
```

ngoldbaum commented:
Cool that works. Thanks for the suggestions, if it isn't clear I know basically nothing about masked arrays :)
